### PR TITLE
Added an example with finding optimal WC hyperparams

### PR DESCRIPTION
**Pull request by l-bat**
### Changes

Added an example to OV backend  with finding optimal parameters (`group_size` and `ratio`) for `INT4_SYM` weight compression.

Merge after #2321 

### Related tickets

127470

### CODE REVIEW

File: examples/weight_compression/openvino/tiny_llama/main.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        that share quantization parameters (scale).
+    :return: The OpenVINO model with compressed weights.
+    """
+    optimized_ov_model = nncf.compress_weights(
+        ov_model.clone(), # we should clone the model because `compress_weights` works in place
+        dataset=nncf_dataset,
+        mode=COMPRESSION_MODE,
+        ratio=ratio,
+        group_size=group_size,
+        sensitivity_metric=nncf.parameters.SensitivityMetric.HESSIAN_INPUT_ACTIVATION,
```
Comment (ljaljushkin) [inline examples/weight_compression/openvino/tiny_llama/main.py:48]: In my latest experiments, MAX_ACTIVATION_VARIANCE gives the best results among other metrics.
Potentially, you can search over them too, maybe in next PR.

File: examples/weight_compression/openvino/tiny_llama/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Weight compression of TinyLLama OpenVINO Model
+
+This example demonstrates how to find optimal ratio and group size to compress weights of the TinyLLama model from the HuggingFace Transformers framework. OpenVINO backend supports mixed precision weight compression with a 4-bit data type as a primary precision. The fastest mixed-precision mode is `INT4_SYM`, but it may lead to a significant accuracy degradation. In this example, the allowed maximum deviation from the original model is `0.2`. If the similarity of the compressed model is not satisfying, there are 2 hyper-parameters to tune: `group_size` and `ratio`. Lower group size and less ratio of 4-bit layers usually improve accuracy at the sacrifice of inference speed.
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/README.md:3]:
```suggestion
This example demonstrates how to find the appropriate `ratio` and `group_size` parameters to compress the weights of the TinyLLama model from the HuggingFace Transformers. OpenVINO backend supports inference of mixed-precision models with weights compressed to a 4-bit data type as a primary precision. The fastest mixed-precision mode is `INT4_SYM`, but it may lead to a significant accuracy degradation, especially for models of moderate size. In this example, the allowed maximum deviation from the original model is `0.2` points of the similarity metric. If the similarity of the compressed model is not satisfying, there are 2 hyper-parameters to tune: `group_size` and `ratio`. Smaller `group_size` and `ratio` of 4-bit layers usually improve accuracy at the sacrifice of model size and inference latency.
```

File: examples/weight_compression/openvino/tiny_llama/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+The example includes the following steps:
+
+- Download and prepare `wikitext` dataset.
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/README.md:7]:
```suggestion
- Prepare `wikitext` dataset.
```

File: examples/weight_compression/openvino/tiny_llama/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- Download and export `PY007/TinyLlama-1.1B-step-50K-105b` model to OpenVINO IR.
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/README.md:8]:
```suggestion
- Prepare `PY007/TinyLlama-1.1B-step-50K-105b` text-generation model in OpenVINO representation using Optimum-Intel.
```

File: examples/weight_compression/openvino/tiny_llama/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+- Compress weights of the model with NNCF Weight compression algorithm.
+- Find optimal ratio and group size if acceptable similarity is not achieved.
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/README.md:10]:
```suggestion
- Find appropriate `ratio` and `group_size` if acceptable similarity is not achieved.
```

File: examples/weight_compression/openvino/tiny_llama/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+- Save model with the best hyperparameters found.
+- Measure similarity and footprint of the compressed model.
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/README.md:12]:
```suggestion
- Measure the similarity and footprint of the final model.
```

File: examples/weight_compression/openvino/tiny_llama/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+## Install requirements
+
+At this point it is assumed that you have already installed NNCF. You can find information on installation NNCF [here](https://github.com/openvinotoolkit/nncf#user-content-installation).
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/README.md:16]: Let's don't do this assumption and consider the user starting from scratch.

File: examples/weight_compression/openvino/tiny_llama/main.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+from nncf.common.logging import nncf_logger
+
+DataItem = TypeVar("DataItem")
+ModelInput = TypeVar("ModelInput")
+
+ROOT = Path(__file__).parent.resolve()
+MODEL_PATH = ROOT / "compressed_model.xml"
+
+COMPRESSION_MODE = nncf.parameters.CompressWeightsMode.INT4_SYM
+MAX_DROP = 0.2
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/main.py:24]: Let's place the comment that we consider the following range of parameters: `group_size` - [64, 128], `ratio` - [0.5,...,1.0]

File: examples/weight_compression/openvino/tiny_llama/main.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    """
+    ov.save_model(optimized_model, MODEL_PATH)
+    footprint = Path(MODEL_PATH).with_suffix(".bin").stat().st_size
+    print(f"Compressed model was saved to: {MODEL_PATH}")
+    print(f"Best parameters: group_size={group_size}, ratio={ratio}")
+    print(f"Memory footprint: {footprint / 2**20 :.2f} MB")
+    print(f"Similarity: {similarity:.2f}")
+
+
+def find_optimal_parameters(evaluator: Evaluator, model: OVModelForCausalLM, nncf_dataset: nncf.Dataset) -> None:
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/main.py:113]:
```suggestion
def find_parameters(evaluator: Evaluator, model: OVModelForCausalLM, nncf_dataset: nncf.Dataset) -> None:
```

File: examples/weight_compression/openvino/tiny_llama/main.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        f"The similarity of model compressed with group_size={group_size}, ratio={ratio} is {similarity:.3f}"
+    )
+    return similarity
+
+
+def get_nncf_dataset(
+    data_source: Iterable[DataItem], transform_func: Optional[Callable[[DataItem], ModelInput]] = None
+) -> nncf.Dataset:
+    """
+    Create an NNCF dataset for the weight compression algorithm.
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/main.py:83]: Just for the future. It is good that you create a description for each function but this is not mandatory for the sample. Without description/typehints, it can make it smaller and even easier to digest.

File: examples/weight_compression/openvino/tiny_llama/main.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    trust_remote_code=True,
+    use_cache=True,
+    ov_config=ov_config,
+)
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+dataset = load_dataset("wikitext", "wikitext-2-v1", split="train[:1000]")
+dataset = dataset.filter(lambda example: len(example["text"]) > 128)
+transform_func = partial(tiny_llama_transform_func, tokenizer=tokenizer)
+
+start = datetime.datetime.now()
```
Comment (AlexKoff88) [inline examples/weight_compression/openvino/tiny_llama/main.py:204]: What is the usual time required to run this sample?

Comment (l-bat) [reply examples/weight_compression/openvino/tiny_llama/main.py:204]: The example requires 45 minutes

File: examples/weight_compression/openvino/tiny_llama/main.py (hunk 1 above)
Comment (l-bat) [reply examples/weight_compression/openvino/tiny_llama/main.py:49]: Thanks for the comment! 
For `HESSIAN_INPUT_ACTIVATION` metric, the appropriate `ratio=0.6` and `group_size=64`, elapsed time - 47 minutes.
For `MAX_ACTIVATION_VARIANCE ` metric, the appropriate `ratio=0.9` and `group_size=128`, elapsed time - 30 minutes.

**Comment by ljaljushkin** [comment]
@l-bat could please make a link to the sample in the https://github.com/openvinotoolkit/nncf/blob/develop/docs/compression_algorithms/CompressWeights.md

File: docs/compression_algorithms/CompressWeights.md [hunk 12]
```diff
@@ -293,6 +293,8 @@ Here is the word perplexity with data-free and data-aware mixed-precision INT4-I
 - [LLM Weight Compression](https://docs.openvino.ai/nightly/weight_compression.html)
 - [Optimize and Deploy Generative AI Models using Hugging Face Optimum Intel](https://docs.openvino.ai/nightly/gen_ai_guide.html)
 - [Optimum Intel documentation](https://huggingface.co/docs/optimum/intel/inference)
+- [Large Language Models Weight Compression Example](https://github.com/openvinotoolkit/nncf/blob/develop/examples/llm_compression/openvino/tiny_llama)
+- [Tuning Ratio and Group Size Example](https://github.com/openvinotoolkit/nncf/blob/develop/examples/llm_compression/openvino/tiny_llama_find_hyperparams)
```
Comment (ljaljushkin) [inline docs/compression_algorithms/CompressWeights.md:297]: Thanks for adding the link to the example! 
I think it would be useful to elaborate more on tuning of hyperparams. For example, after words:
>If the accuracy or perplexity is still not satisfying, there are 2 more hyper-parameters to tune: `group_size` and `ratio`.

File: examples/llm_compression/openvino/tiny_llama_find_hyperparams/README.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
+# Find the appropriate hyperparameters to compress the TinyLLama model
+
+This example demonstrates how to find the appropriate `ratio` and `group_size` parameters to compress the weights of the TinyLLama model from the HuggingFace Transformers. OpenVINO backend supports inference of mixed-precision models with weights compressed to a 4-bit data type as a primary precision. The fastest mixed-precision mode is `INT4_SYM`, but it may lead to a significant accuracy degradation, especially for models of moderate size. In this example, the allowed maximum deviation from the original model is `0.2` points of the similarity metric. If the similarity of the compressed model is not satisfying, there are 2 hyper-parameters to tune: `group_size` and `ratio`. Smaller `group_size` and `ratio` of 4-bit layers usually improve accuracy at the sacrifice of model size and inference latency.
```
Comment (AlexKoff88) [inline examples/llm_compression/openvino/tiny_llama_find_hyperparams/README.md:3]: I think you should mention WhoWhatBench project here as well, like. "To evaluate the accuracy of the compressed model we measure similarity between two texts generated by the baseline and compressed models using [WhoWhatBench](link here) library."

File: examples/llm_compression/openvino/tiny_llama_find_hyperparams/README.md [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+The example includes the following steps:
+
+- Prepare `wikitext` dataset.
+- Prepare `TinyLlama/TinyLlama-1.1B-step-50K-105b` text-generation model in OpenVINO representation using Optimum-Intel.
```
Comment (AlexKoff88) [inline examples/llm_compression/openvino/tiny_llama_find_hyperparams/README.md:8]: Please provide a links to Optimum-Intel

File: examples/llm_compression/openvino/tiny_llama_find_hyperparams/README.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+```
+
+## Run Example
+
+The example is fully automated. Just run the following command in the prepared Python environment:
+
+```bash
+python main.py
+```
+
```
Comment (AlexKoff88) [inline examples/llm_compression/openvino/tiny_llama_find_hyperparams/README.md:33]: Since this is an example it would be helpful to explain how to plug a different model model. Maybe it would be better to place comments about this in the code as well.

File: examples/llm_compression/openvino/tiny_llama_find_hyperparams/main.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    Find the optimal `ratio` and `group_size` for weight compression algorithm.
+
+    :param evaluator: The evaluator object from whowhatbench Benchmark.
+    :param model: The OpenVINO model for causal language modeling.
+    :param nncf_dataset: A representative dataset for the weight compression algorithm.
+    """
+    original_ov_model = model.model
+    evaluate_fn = partial(evaluate_model, hf_model=model, original_ov_model=original_ov_model, evaluator=evaluator)
+
+    param_grid = list(itertools.product(list((x / 10 for x in range(10, 4, -1))), [MAX_GROUP_SIZE, MIN_GROUP_SIZE]))
```
Comment (AlexKoff88) [inline examples/llm_compression/openvino/tiny_llama_find_hyperparams/main.py:136]: Some inline comments about the major steps would be helpful in this function.

File: docs/compression_algorithms/CompressWeights.md (hunk 12 above)
Comment (ljaljushkin) [reply docs/compression_algorithms/CompressWeights.md:300]: Now it looks out of the context. 

I meant a different thing. This sentence is already in the section with example. Please mention your script after that sentence there with some quick overview. e.g.
```
If the accuracy or perplexity is still not satisfying, there are 2 more hyper-parameters to tune: `group_size` and `ratio`. Please refer to the [example](link) how to automatically tune these parameters. 

```

**Review by ljaljushkin — APPROVED** [review]

**Comment by AlexKoff88** [comment]
LGTM
